docs/apache-airflow/concepts/scheduler.rst

```diff
@@ -138,18 +141,173 @@ The following databases are fully supported and provide an "optimal" experience:
+* Another solution to FileSystem performance, if it becomes your bottleneck, is to turn to alternative
+  mechanisms of distributing your DAGs. Embedding DAGs in your image and GitSync distribution have both
+  the property that the files are available locally for Scheduler and it does not have to use a
+  distributed filesystem to read the files, the files are available locally for the Scheduler and it is
+  usually as fast as it can be, especially if your machines use fast SSD disks for local storage. Those
+  distribution mechanisms have other characteristics that might make them not the best choice for you,
+  but if your problems with performance come from distributed filesystem performance, they might be the
+  best approach to follow.
+* Database connections and Database usage might become a problem as you want to increase performance and
+  process more things in parallel. Airflow is known from being "database-connection hungry" - the more DAGs
+  you have and the more you want to process in parallel, the more database connections will be opened.
+  This is generally not a problem for MySQL as its model of handling connections is thread-based, but this
+  might be a problem for Postgres, where connection handling is process-based. It is a general consensus
+  that if you have even medium size Postgres-based Airflow installation, the best solution is to use
+  `PGBouncer <https://www.pgbouncer.org/>`_ as a proxy to your database. The :doc:`helm-chart:index`
+  supports PGBouncer out-of-the-box. For MsSQL we have not yet worked out the best practices as support
+  for MsSQL is still experimental.
+* CPU usage is most important for FileProcessors - those are the processes that parse and execute
+  Python DAG files. Since Schedulers triggers such parsing continuously, when you have a lot of DAGs,
+  the processing might take a lot of CPU. You can mitigate it by decreasing the
+  :ref:`config:scheduler__min_file_process_interval`, but this is one of the mentioned trade-offs,
+  result of this is that changes to such files will be picked up slower and you will see delays between
+  submitting the files and getting them available in Airflow UI and executed by Scheduler. Optimizing
+  the way how your DAGs are built, avoiding external data sources is your best approach to improve CPU
+  usage. If you have more CPUs available, you can increase number of processing threads
+  :ref:`config:scheduler__parsing_processes`, Also Airflow Scheduler scales almost linearly with
+  several instances, so you can also add more Schedulers if your Scheduler's performance is CPU-bound.
+* Airflow might use quite significant amount of memory when you try to get more performance out of it.
+  Often more performance is achieved in Airflow by increasing number of processes handling the load,
+  and each process requires whole interpreter of Python loaded, a lot of classes imported, temporary
```

Comment:
Because of forking and copy-on-write, each process doesn't pay the full cost of this memory usage: it is shared between processes,

i.e. one process might have 300Mb of active memory, but a second process doesn't need an extra 300Mb, it needs _almost_ nothing extra apart from the data it processes in memory.

Reply:
Agree in general for all the "shared C libraries" and "shared python libraries loaded before fork" part. I will reword it to explain it better. It's indeed all about Heap Space and Stack under the hood, but it also depends a bit on a sequence of loading - for example with the "Optimized Hook imports" on one hand you have faster startup, but on the other hand, the lirbraries that are imported after the fork are duplicated - so the truth is somewhere in-between.